### PR TITLE
[Serialization] Delay adding ModuleFiles until they're loaded

### PR DESCRIPTION
**Pull request by hamishknight**
Rather than adding a `ModuleFile` to a parent module and then removing it afterwards if it fails to load, let's wait until we've loaded the file before deciding to add it to the parent module. This then allows us to get rid of `ModuleDecl::removeFile`.

In addition, push down the call to `addFile` into the callers of `loadAST` in preparation for `addFile` being replaced with a one-time-only call to a `setFiles` method.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by harlanhaskins** [comment]
Can you add an assertion in `addFile` that the file has loaded? Just to make sure we don't violate this invariant later.

File: lib/AST/Module.cpp [hunk 1]
```diff
@@ -507,19 +507,6 @@ void ModuleDecl::addFile(FileUnit &newFile) {
...
-  // Do a reverse search; usually the file to be deleted will be at the end.
-  std::reverse_iterator<decltype(Files)::iterator> I(Files.end()),
-  E(Files.begin());
-  I = std::find(I, E, &existingFile);
-  assert(I != E);
-
-  // Adjust for the std::reverse_iterator offset.
-  ++I;
-  Files.erase(I.base());
-  clearLookupCache();
```
Comment (slavapestov) [inline lib/AST/Module.cpp:520]: The only remaining call to clearLookupCache() is from addFile(). Is that one still necessary?

**Review by slavapestov — APPROVED** [review]
Awesome!

File: lib/AST/Module.cpp (hunk 1 above)
Comment (hamishknight) [reply lib/AST/Module.cpp:520]: @slavapestov I *think* the Tensorflow folk are currently relying on it to clear the module-level cache for when they add a `SynthesizedFileUnit` to the module. I'm hoping to take care of that when I replace `addFile` with `setFiles`.

Comment (slavapestov) [reply lib/AST/Module.cpp:520]: Ah yeah, I just saw that in your PR description. Makes sense, thanks!

**Review by xymus — APPROVED** [review]
Great! I wonder if this will not fix a few crashes we have iterating through loaded files in the middle of loading files.

**Comment by hamishknight** [comment]
@harlanhaskins Good idea! How does this look?

**Comment by hamishknight** [comment]
@swift-ci please test

**Review by harlanhaskins — APPROVED** [review]

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by harlanhaskins** [comment]
Thanks!

**Comment by hamishknight** [comment]
@swift-ci please test Windows

**Comment by hamishknight** [comment]
@swift-ci please test Windows

**Comment by hamishknight** [comment]
@compnerd Looks like the Windows bot is out of disk space 

```
fatal error C1088: Cannot flush compiler intermediate file: 'C:\Users\ci\AppData\Local\Temp\_CL_01aae7eaex': No space left on device
```

**Comment by compnerd** [comment]
I believe that I’ve cleaned them up again.

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

**Pull request merged**